kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -484,6 +484,178 @@ sealed abstract class Resource[F[_], +A] {
     new (F ~> F) {
       override def apply[B](gb: F[B]): F[B] = surround(gb)
     }
+
+  def forceR[B](that: Resource[F, B])(implicit F: MonadCancel[F, Throwable]): Resource[F, B] =
+    Resource applyFull { poll =>
+      poll(this.use_ !> that.allocated) map { p =>
+        // map exists on tuple in Scala 3 and has the wrong type
+        Functor[(B, *)].map(p)(fin => (_: Resource.ExitCase) => fin)
+      }
+    }
+
+  def onCancel[B >: A](fin: Resource[F, Unit])(
+      implicit F: MonadCancel[F, Throwable]): Resource[F, B] =
```

Comment:
I think this one *shouldn't* need the context bounds.

```suggestion
  def onCancel(fin: Resource[F, Unit])(
      implicit F: MonadCancel[F, Throwable]): Resource[F, A] =
```